src/stmt.rs

```diff
@@ -80,13 +80,19 @@ impl<'a> Rewrite for Stmt<'a> {
         } else {
             ExprType::Statement
         };
-        format_stmt(context, shape, self.as_ast_node(), expr_type)
+        format_stmt(
+            context,
+            shape,
+            self.as_ast_node(),
+            expr_type,
+            Some(self.is_last_expr()),
+        )
     }
 }
 
 impl Rewrite for ast::Stmt {
     fn rewrite(&self, context: &RewriteContext<'_>, shape: Shape) -> Option<String> {
-        format_stmt(context, shape, self, ExprType::Statement)
+        format_stmt(context, shape, self, ExprType::Statement, None)
     }
 }
```

Comment:
Can you check which places use this `Rewrite` impl? You could probably look at the errors if you remove this impl.

Reply:
Good idea! It looks like the three places that it gets used are (lines 520, 811, and 815)

https://github.com/rust-lang/rustfmt/blob/9386b32f5a72bb276a97461d33aab415314201d8/src/expr.rs#L518-L520

https://github.com/rust-lang/rustfmt/blob/9386b32f5a72bb276a97461d33aab415314201d8/src/expr.rs#L802-L815

All three of those places currently gate on the containing block being a `is_simple_block()` (only an expr with no statements), so that wouldn't be true in cases that cause this issue. I doubt that that will always hold though, so it may be worth dropping `impl Rewrite for ast::Stmt` entirely for `crate::stmt::Stmt` since the only additional information needed is `is_last` which is known for all of those

Reply:
Perhaps instead of dropping the impl and having to update all the call sites to do the same things (e.g. create a stmt::Stmt) the rewrite could be updated to facilitate that translation?

Reply:
Any idea for how you would want that to look? I'm not sure of a way to handle it in a way that doesn't clash with rewrite for `crate::stmt::Stmt` already containing that information

Reply:
Ah yes, I suppose it might be difficult for callers to indicate whether it's the last statement through a `rewrite` call wouldn't it :laughing:.

I don't recall offhand if there's already a constructor function for the stmt::Stmt, but if not, would be good to add something that will encapsulate the creation of the stmt::Stmt wrapper and invokes `format_stmt`

Reply:
I've settled on adding a `stmt::Stmt::from_simple_block(...) -> Option<Self>` constructor to switch all the existing `ast::Stmt` rewriting over to

The only reason I didn't go for adding in a helper that also invokes `format_stmt` is because there is shaping that requires additional context between checking that it was a simple block and rewriting it, and I didn't want to have to switch to shaping before and then fixing up shaping on failures

Let me know what ya think

(aside: is this one of the repos where it's generally expected to rebase before the final merge?)

Reply:
> (aside: is this one of the repos where it's generally expected to rebase before the final merge?)

I don't think you'd need to rebase on top of master before merging, but it would be nice if you could avoid having merge commits in your PR.

Reply:
Fair enough. The commit history should be clean now